### DISCUSSION TITLE
Possible GSOC project: RDKit DuckDB Extension

### DISCUSSION THREAD

**Discussion by apahl** [rdkit/rdkit#7097] [Ideas]
[DuckDB](https://duckdb.org/) is a cross-platform serverless SQL database engine that is gaining a lot of traction. It is similar to SQLite in that it can be easily installed and uses a single file approach for the database, but in addition it offers a lot of features that are appealing to data scientists, like (quoting from the landing page):

- Vectorized and parallel engine
- Larger than memory processing
- Direct Parquet, CSV, and JSON querying
- Free & open-source
- Permissive MIT License
- Flexible extension mechanism

I have no idea how much work it would be, but it might be a cool Google Summer of Code (GSOC) project to generate an RDKit extension for DuckDB (DuckDB is written in C++).

KR Axel

**Comment by ghost** [comment]
I'm also interested in this idea and started exploring how to add an extension to duckdb. 
I've been looking at the spatial extension from duckdb, as well as chemicalite (the sqlite rdkit extension) for examples.
I haven't gotten very far. So far I can get rdkit and duckdb compiled, and I was able to add a molecule type and some basic mol_from_smiles working.
But so far, I'm not able to get mol_to_smiles when pulling the pickled mol out of a table. 

While I'm at it, I am getting an `Bad Pickle format: bad endian ID or invalid file format` error when trying to convert mol back to smiles when I run RDKit's molFromPickle. 
I'm still trying to figure out if I did something incorrectly with RDKit or if the error is due to something added by duckdb. 
If anyone can point me in the right direction, please let me know. Thank you!
